### PR TITLE
docs: update from internal docs

### PR DESCRIPTION
**Pull request by mvayngrib**
- pull in some updates from https://docs.google.com/document/d/1kkOtkFM-vw0UVuARiYv2tq4FzF80vq48Ol_uK28yHs0/edit
- re-weigh the points
- change some points to negative-only - i.e. candidate can only lose points for useless comments, they don't gain points if they don't have any useless ones

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,25 +1,74 @@
-# Code Guidelines
-For reference, see the below guidelines we use for peer code review at Exodus.
+# Code and Process Guidelines
+
+For reference, see the below guidelines we use for peer code review at Exodus. You can follow this checklist when evaluating a code challenge submission.
+
+## Grading
+
+See points in parentheses before each item. Negative points means the candidate can only lose points for that item. The expected Total Score:
```
Comment (timlanahan) [inline README.md]:
```suggestion
See points in parentheses before each item. Negative points mean the candidate can only lose points for that item. The expected Total Score:
```

File: README.md [hunk 2]
```diff
@@ -1,25 +1,74 @@
...
+
+- Candidates should score a total of >15 points if they use PRs, >10 otherwise
+- Colleagues should score a total of >20 points
```
Comment (timlanahan) [inline README.md]: Do you think we should just strip out the `Colleagues` part of this doc and simply make it about candidates?  If we want to acknowledge that this is how we do things internally, I think the way you called it out at the beginning is good.  Idk if candidates need to know that we should be >20 (esp b/c idk we are rating each other this way 😁 )

File: README.md [hunk 3]
```diff
@@ -1,25 +1,74 @@
...
+
+## Code
+
+> [!IMPORTANT]
+> GRADING (SECTION SCORE)
+>
+> - Everyone should score at least 10 points (tests optional for candidates)
```
Comment (timlanahan) [inline README.md]: What do you think about getting rid of the "optional" piece and leaving it to the imagination?

```suggestion
> - Everyone should score at least 10 points
```

File: README.md [hunk 4]
```diff
@@ -1,25 +1,74 @@
...
+
+### Functionality
+
+- [ ] (1) All task requirements are fully implemented, tradeoffs made are documented, explanations are provided for any missing pieces
```
Comment (timlanahan) [inline README.md]: Do we allow for fractions of a point in case a single point is a compound statement made up of a few things?

File: README.md [hunk 5]
```diff
@@ -1,25 +1,74 @@
...
+- [ ] (1) Not obviously broken
+- [ ] (1) No dubious functionality on top of the spec
+- [ ] (1) Not over-engineered, e.g., doesn’t prematurely optimize at the expense of complexity
+
+### Code Quality
+
+- [ ] (2) Code is readable: linted, prettified, and written in a consistent style
+- [ ] (1) Code is self-documenting: semantic variables names, function names describe implemented function
+- [ ] (1) Code is as simple as possible and modular
+  - Small single-purpose functions > giant functions that do 10 things
```
Comment (timlanahan) [inline README.md]:
```suggestion
  - Small single-purpose functions are preferred to giant functions that do 10 things
```

File: README.md [hunk 6]
```diff
@@ -1,25 +1,74 @@
...
+  - Code is logically separated (e.g., no giant utils folders)
+- [ ] (2) Author uses objects, arrays, inheritance, closures, callbacks, promises expertly, and is up to date with the language/technology, e.g., async/- [ ] await > promise chains > callbacks for code after 2019
+- [ ] (1) Tests are included and:
```
Comment (timlanahan) [inline README.md:31]: Should tests be greater than a single point?  if someone actually writes tests and they're really good, maybe we should consider this a two pointer?

File: README.md [hunk 7]
```diff
@@ -1,25 +1,74 @@
...
+  - Are readable (intent is clear)
+  - Cover happy path
+  - Cover edge cases
+
+### Documentation
+
+- [ ] (2) Clear instructions for project setup
+- [ ] (-1) Clear code comments, but _only when necessary_. Comments are for the why, not the what
```
Comment (timlanahan) [inline README.md:39]: What if we start a "Negatives" section and breakout all of the offensive code smells there?

File: README.md [hunk 8]
```diff
@@ -1,25 +1,74 @@
...
-    - Cover edge cases
\ No newline at end of file
+
+> [!IMPORTANT]
+>
+> - Candidates should score at least 5 points in this section (or skip if not working with Pull Requests, e.g. in the context of a code challenge)
+> - Colleagues should score at least 10 points in this section
+
+- [ ] (1) Commits tend towards atomic (take care of a single thing)
+- [ ] (1) Commit messages are descriptive and follow the [seven rules of a great Git commit message](https://chris.beams.io/posts/git-commit/)
```
Comment (timlanahan) [inline README.md:44]: These are both under the section that says to skip if we're not doing PRs...perhaps we should have a section called `Git` and stick these there.  Then we should move on to PRs in a single section.

File: README.md [hunk 9]
```diff
@@ -1,25 +1,74 @@
...
+
+If using Pull Requests:
+
+- [ ] (1) Pull requests are atomic (take care of a single thing) and functional (don’t leave the program in a half-functioning state)
+- [ ] (1) Pull requests have proper description, mentioning the what and the why and (infrequently and if complex) the how, and include detailed test plans (steps + expected outcome)
+- [ ] (1) Knows how to rebase commits
```
Comment (timlanahan) [inline README.md]: But how do we judge this?  I'm guessing someone working solo dev is not going to have the opportunity to prove this except if they force a scenario for themselves.

File: README.md [hunk 10]
```diff
@@ -1,25 +1,74 @@
...
+- [ ] (1) Knows how and when to squash/fixup commits
```
Comment (timlanahan) [inline README.md]: same -- idk if we can judge this for a candidate

File: README.md [hunk 11]
```diff
@@ -1,25 +1,74 @@
...
+- [ ] (1) Pull requests follow the Pull Request template in the repo (if one is present)
```
Comment (timlanahan) [inline README.md]: Should we add a PR template to our code challenge repos?

File: README.md [hunk 12]
```diff
@@ -1,25 +1,74 @@
...
+- [ ] (1) Knows how to create and maintain (through rebasing) a chain of related pull requests for easier review and merging (Bonus for candidates, requirement for colleagues)
+
+## Code Reviews
```
Comment (timlanahan) [inline README.md:52]: Do we think this will ever be relevant for candidates?

File: README.md (hunk 2 above)
Comment (mvayngrib) [reply README.md]: sgtm

File: README.md (hunk 4 above)
Comment (mvayngrib) [reply README.md]: yeah, we can split this up into 3, e.g. 1 point for fully implemented, 0.5 for documentation of missing pieces/tradeoffs

File: README.md (hunk 6 above)
Comment (mvayngrib) [reply README.md:31]: maybe we should give ourselves the freedom to give an extra point for any section that's done really well? (candidate reviewer's discretion?)

File: README.md (hunk 7 above)
Comment (mvayngrib) [reply README.md:39]: we could, it just depends on how short we want to keep this. Or we can have a separate phase to the challenge, e.g. please stop after 7 hours and spend the last hour working on refactoring towards our coding best practices

File: README.md (hunk 8 above)
Comment (mvayngrib) [reply README.md:44]: sgtm 👍

File: README.md (hunk 9 above)
Comment (mvayngrib) [reply README.md]: we don't, this is for judging colleagues. We can remove it or move it to a separate doc

File: README.md (hunk 11 above)
Comment (mvayngrib) [reply README.md]: if we want people to actually use them. This is more for colleagues i guess

File: README.md (hunk 12 above)
Comment (mvayngrib) [reply README.md:52]: it would be cool!

File: README.md (hunk 9 above)
Comment (mvayngrib) [reply README.md]: removed

File: README.md (hunk 11 above)
Comment (mvayngrib) [reply README.md]: made this a -1 instead

File: README.md (hunk 8 above)
Comment (mvayngrib) [reply README.md:44]: 6a96b45edece182d62360cd5bbb4447c5a461758

File: README.md (hunk 4 above)
Comment (mvayngrib) [reply README.md]: 28f1f78c3ea3e40cea1fc3ffda9f5cec26f4032f

**Comment by mvayngrib** [comment]
added a bonus point 8e2db070084d120385988121e2f5838e15360780

File: README.md (hunk 6 above)
Comment (timlanahan) [reply README.md:31]: Perhaps if someone is on the fence and you just have that gut feeling that so-and-so is gonna be good.  But that might apply for all sections wrt "discretion"

File: README.md (hunk 7 above)
Comment (timlanahan) [reply README.md:39]: Let's keep it like you have it then.  I always assume that most ppl spend more than 8 hours so any time refactoring will be spent by someone who is determined to do well anyway

File: README.md (hunk 12 above)
Comment (timlanahan) [reply README.md:52]: Actually we could do that.  We are talking about our expedited challenge but what if we went the other way and said "you have 5 days to implement X.  Please make a PR for these features" -- might be too complicated though especially because we'd have to be on the ball with reviews

**Review by timlanahan — APPROVED** [review]
utACK

File: README.md [hunk 13]
```diff
@@ -1,25 +1,63 @@
...
+  - Code is logically separated (e.g., no giant utils folders)
+- [ ] (2) Author uses objects, arrays, inheritance, closures, callbacks, promises expertly, and is up to date with the language/technology, e.g., async/- [ ] await > promise chains > callbacks for code after 2019
+- [ ] (1) Tests are included and:
+  - Are readable (intent is clear)
+  - Cover happy path
+  - Cover edge cases
+
+### Documentation
+
+- [ ] (2) Clear instructions for project setup
```
Comment (mvayngrib) [inline README.md]: 2 is a bit too much 🤔. I'm thinking either 1 or -1, @timlanahan ?
```suggestion
- [ ] (-1) Clear instructions for project setup
```

Comment (timlanahan) [reply README.md]: totally agree...this should be a baseline.  maybe negative point is best (which i think is fair esp if it's in the guidelines).  at the very least someone should be able to write a basic description of how to install deps and run the project.  i don't think that should take too long

**Review by timlanahan — APPROVED** [review]
utACK

**Review by diegomura — APPROVED** [review]
utACK

**Review by andrejborstnik — APPROVED** [review]
utACK

**Review by olistic — APPROVED** [review]
